### ISSUE TITLE
Call the macos file dialog from Main thread only

### ISSUE THREAD

**Issue by AChep** [btzy/nativefiledialog-extended#46]
The latest macos versions will crash if you try to call its UI components from a non-main thread. Therefore it would be nice if we dispatch all of the calls to the main thread.

**Comment by btzy** [comment]
I'm thinking this might be more of the responsibility of the user instead of the library.  Most UI libraries assume that UI updates are only invoked on the main thread.

As far as I know, both GTK and Windows require all UI updates to be from one specific thread only.  Some GTK functions may be called by other threads, but you cannot make a UI update from those threads.  So if we want your behaviour for MacOS, we ought to implement it for Windows and Linux too.

But I think we should not provide this feature at all - most people do not need such a feature, and it complicates the implementations.  Currently, the calling thread gets blocked until the file is selected.  But if the API is meant to be called from multiple threads, we will need to determine what we should do when multiple threads call it simultaneously, and ensure that it can be implemented consistently across all platforms - I think this will be over-complicated or impossible.

Those who want such a feature can easily implement the synchronisation on their own - just dispatch the call on your own, and invoke NFDe inside the callback that runs on the main thread.

**Comment by AChep** [comment]
Makes sense, even if it hurts my soul knowing that I can not get the main thread from Java -- therefore needed another native layer that dispatches the stuff. :(

**Comment by btzy** [comment]
Hmm, I thought you already have to write some wrapper C code to invoke anything from Java - is this not the case?  Could you give me a link to your project where you use this (if it's on GitHub)?

**Comment by AChep** [comment]
I can not give you the full source code, but we Java developers have this little library called Java Native Access that helps with that issue.

An example:
```
interface NfdLibraryPure {
    /* initialize NFD - call this for every thread that might use NFD, before calling any other NFD
     * functions on that thread */
    fun NFD_Init(
    ): NfdResult

    /* call this to de-initialize NFD, if NFD_Init returned NFD_OKAY */
    fun NFD_Quit()

    /*
     * Select folder dialog
     *
     * It is the caller's responsibility to free `outPath` via NFD_FreePathN() if
     * this function returns NFD_OKAY.
     *
     * If defaultPath is NULL, the operating system will decide
     */
    fun NFD_PickFolderN(outPath: PointerByReference, defaultPath: WString?): NfdResult

    fun NFD_FreePathN(filePath: PointerByReference)
}
```

**Comment by AChep** [comment]
it basically a C wrapper in disguise :smile:

**Comment by btzy** [comment]
I see, interesting :)

**Comment by AChep** [comment]
I guess if I'll have some free time, I will create a Java wrapper library for this one :) For JetPack Compose it is gonna be a perfect match.

**Comment by btzy** [comment]
Thanks!  Maybe you could also take a look at the original NFD (that this project was forked from), I think it has some kind of Java bindings (https://javadoc.lwjgl.org/org/lwjgl/util/nfd/NativeFileDialog.html).  But I think they also don't have support for calling the function via a non-main thread.
